core-bundle/src/Resources/contao/templates/image/image.html5

```diff
@@ -14,3 +14,9 @@
     <figcaption class="caption"><?= $this->caption ?></figcaption>
   <?php endif; ?>
 </figure>
+
+<?php
+
+if ($this->figure) {
+    $this->jsonLd($this->figure->getJsonLd());
+}
```

Comment:
Personally I would prefer having

```php
<?php

if ($this->figure) {
    $this->jsonLd($this->figure->getJsonLd());
}

?>
```

at the beginning of the template, but since we do not really have a convention for that (yet) I am fine with this. Only question remains: since templates use two spaces for indentation, should PHP code within `<?php ?>` in templates also use two spaces or four spaces like here?

@m-vo how would the implementation look in an `image.html.twig`?

Reply:
I prefer it to be at the end because it's meta data. and the indendation is set to 4 by the cs fixer :)

Reply:
There is the `do` tag in Twig to execute things. We would need a `jsonLd` template function and then write this:

```twig
{% do jsonLd(figure.jsonLd) %}
```

@toflar We should IMO add the Twig function to this PR as well. I can contribute this if you want to.

Reply:
No, I don't want any Twig here please. There's a gazillion other places where we will be able to add JSON-LD and make use of the new functionality. Those are future PR's. Feel free to create one as soon as this is merged :)

Reply:
> I prefer it to be at the end because it's meta data. and the indendation is set to 4 by the cs fixer :)

Ah ok, didn't realise the CS fixer already covers this :)

Reply:
> There's a gazillion other places where we will be able to add JSON-LD.

Then this should imo become a service and not live inside `Template`.

Reply:
Yes, which it already is. It's a one-liner.

Reply:
I would also prefer to have this at the beginning of the template, because that is where we usually add metadata such as `$GLOBALS['TL_CSS']` etc. @Toflar Mind if I change this?

Reply:
The reason I put it at the bottom is because it might get a lot longer than just one line in other templates. With various `if` statments. Dunno if you remember the adjusted `news_full` template but it looked (and will look as I will provide PR's that actually use the new feature in other places soon) something like this:

```php
<?php

$jsonLd = [
    '@type' => 'NewsArticle',
    'identifier' => $this->absoluteLink,
    'url' => $this->absoluteLink,
    'headline' => $this->raw($this->headline),
    'datePublished' => $this->datetime,
];

if ($this->hasTeaser) {
    $jsonLd['description'] = $this->raw($this->teaser);
}

if ($this->hasText()) {
    $jsonLd['text'] = $this->raw($this->text);
}

if ($this->addImage) {
    $jsonLd['image'] = $this->figure->getJsonLd();
}

if ($this->authorModel) {
    $jsonLd['author'] = [
        '@type' => 'Person',
        'name' => $this->authorModel->name,
    ];
}

$this->jsonLd($jsonLd);
```

So you'd want to have this whole block bevore any HTML? Because for consistency reasons, we should handle all templates the same.

Reply:
It could be shortened to

```php
<?php

$this->jsonLd(array_filter([
    '@type' => 'NewsArticle',
    'identifier' => $this->absoluteLink,
    'url' => $this->absoluteLink,
    'headline' => $this->raw($this->headline),
    'datePublished' => $this->datetime,
    'description' => $this->hasTeaser ? $this->raw($this->teaser) : null,
    'text' => $this->hasText() ? $this->raw($this->text) : null,
    'image' => $this->addImage ? $this->figure->getJsonLd() : null,
    'author' => $this->authorModel ? ['@type' => 'Person', 'name' => $this->authorModel->name] : null,
]));
```

but I see your point.

Reply:
I know but remember that regular users are editing templates. That's why I would not use `array_filter()` stuff here. It sould be as simple as it can get so everybody understands what it does.

Reply:
Is it unavoidable to have this code in a template? Or could we encapsulate it in a helper class or something alike?

Reply:
It's there so that you can easily change or extend it by altering the template - just as you would be able to before the JSON-LD management by editing the HTML tags and attributes.

Reply:
We may sure think about also providing a similar helper like we did on `Figure` in this PR on `NewsModel` so the basic data can be reused in a Twig template as well. However, the whole purpose of having the `addSchemaOrg()` method in the template is for it to be fully customizable. So the idea really is that you can modify the array to your personal needs and this might easily exceed a few lines of code. That's why I placed it at the end and I still prefer this. The primary use case of the template is to manage the HTML. That's gonna be the 90% use case. So HTML should be first in the template.

Reply:
Maybe it would be easier to understand in templates if the method would (additionally) allow to define the entries one by one?

```php
$this->jsonLd('@type', 'NewsArticle');
$this->jsonLd('url', $this->absoluteLink);
```

```twig
{% do addJsonLd('@type', 'NewsArticle') %}
{% do addJsonLd('url', absoluteLink) %}
```

Reply:
Another thing I was wondering: How is inheritance affected by this? In a child template everything outside blocks is ignored, so you could only alter JSON-LD data inside any block. It could neither be at the top or bottom of the document then.

Workaround would be to always put the call into blocks in the first place.

Reply:
> In a child template everything outside blocks is ignored

Only after `$this->extend()` - so any PHP code before that is still evaluated.

Reply:
That's not the case in Twig, though and trying to support it would be unnecessarily complex.

Reply:
Just checked: `do` works outside blocks as it does not emit any output.